ssl/ssl_test.cc

```diff
@@ -4422,6 +4429,77 @@ static bool ExpectSingleError(int lib, int reason) {
+
+  ASSERT_TRUE(SSL_CTX_use_certificate(ctx.get(), GetLeafPublic().get()));
+  ASSERT_TRUE(SSL_CTX_use_PrivateKey(ctx.get(), GetLeafKey().get()));
+
+  // Verification will fail because there is no valid root cert available.
+  EXPECT_FALSE(SSL_CTX_build_cert_chain(ctx.get(), 0));
+
+  // Should return 2 when |SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR| is set.
+  EXPECT_EQ(
+      SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR),
+      2);
+  EXPECT_TRUE(ExpectSingleError(ERR_LIB_SSL, SSL_R_CERTIFICATE_VERIFY_FAILED));
+
+  // Should return 2, but with no error on the stack when
+  // |SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR| and |SSL_BUILD_CHAIN_FLAG_CLEAR_ERROR|
+  // are set.
+  EXPECT_EQ(
+      SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR |
+                                              SSL_BUILD_CHAIN_FLAG_CLEAR_ERROR),
+      2);
+  EXPECT_FALSE(ERR_get_error());
+
+  // Pass in the trust store. |SSL_CTX_build_cert_chain| should succeed now.
+  ASSERT_TRUE(X509_STORE_add_cert(SSL_CTX_get_cert_store(ctx.get()),
+                                  GetLeafRoot().get()));
+  X509_VERIFY_PARAM_set_flags(SSL_CTX_get0_param(ctx.get()),
+                              X509_V_FLAG_NO_CHECK_TIME);
+  EXPECT_EQ(SSL_CTX_build_cert_chain(ctx.get(), 0), 1);
+  STACK_OF(X509) *chain;
+  ASSERT_TRUE(SSL_CTX_get0_chain_certs(ctx.get(), &chain));
+  EXPECT_TRUE(ChainsEqual(chain, {GetLeafRoot().get()}));
+
+  // Root cert is self-signed, so |SSL_BUILD_CHAIN_FLAG_UNTRUSTED| will
+  // still pass.
+  ASSERT_TRUE(SSL_CTX_clear_chain_certs(ctx.get()));
+  EXPECT_TRUE(
+      SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_UNTRUSTED));
+  ASSERT_TRUE(SSL_CTX_get0_chain_certs(ctx.get(), &chain));
+  EXPECT_TRUE(ChainsEqual(chain, {GetLeafRoot().get()}));
+
+  // |SSL_BUILD_CHAIN_FLAG_CHECK| uses the already built cert chain as the trust
+  // store and verifies against it. If we clear the cert chain, there should be
+  // no trust store to compare against if |SSL_BUILD_CHAIN_FLAG_CHECK| is still
+  // set.
+  EXPECT_EQ(SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_CHECK), 1);
+  ASSERT_TRUE(SSL_CTX_clear_chain_certs(ctx.get()));
+  EXPECT_FALSE(SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_CHECK));
+  EXPECT_TRUE(ExpectSingleError(ERR_LIB_SSL, SSL_R_CERTIFICATE_VERIFY_FAILED));
+
+  // Test that successful verification with |SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR|
+  // does not return 2.
+  ASSERT_TRUE(SSL_CTX_clear_chain_certs(ctx.get()));
+  EXPECT_EQ(
+      SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR),
+      1);
+
+  // Test that successful verification with |SSL_BUILD_CHAIN_FLAG_NO_ROOT|
+  // does include the root cert.
+  ASSERT_TRUE(SSL_CTX_clear_chain_certs(ctx.get()));
+  EXPECT_EQ(SSL_CTX_build_cert_chain(ctx.get(), SSL_BUILD_CHAIN_FLAG_NO_ROOT),
+            1);
+  ASSERT_TRUE(SSL_CTX_get0_chain_certs(ctx.get(), &chain));
+  EXPECT_TRUE(ChainsEqual(chain, {}));
+}
```

Comment:
Can only one flag be passed into this function (if so should we add code to enforce that)? Or should we test all 2^5 combinations of flags?

Reply:
* I don't think there are any restrictions. I'd be careful with adding our own code to enforce things though, especially if the goal is to gain more compatibility with OpenSSL.
* As for testing all 2^5 combinations, most of these flag options don't exactly interfere with one another so I'm not sure if there's any additional value in testing every single combination. For example, `..._NO_ROOT` is just removing the root certificate from the chain, while `..._IGNORE_ERROR` and `..._CLEAR_ERROR` are only handling errors. I'll add a test for the mutual exclusiveness of `..._UNTRUSTED` and `..._FLAG_CHECK` though.